valhalla/baldr/traffictile.h

```diff
@@ -0,0 +1,190 @@
+  uint64_t tile_id;
+  uint32_t directed_edge_count;
+  uint32_t incident_buffer_size;
+  uint64_t active_incident_buffer : 1;
+  uint64_t last_update : 63; // seconds since epoch
+};
+
+#ifndef C_ONLY_INTERFACE
+// Some checks to ensure that the interfaces don't get change accidentally.
+// Modifying the sizes/layouts of these structs is a data-format breaking
+// change and shouldn't be done lightly.
+static_assert(sizeof(Incident) == sizeof(uint64_t),
+              "traffic:Incident type size is different than expected");
+static_assert(sizeof(TileHeader) == sizeof(uint64_t) * 3,
+              "traffic:TileHeader type size different than expected");
+static_assert(sizeof(Speed) == sizeof(uint16_t),
+              "traffic::Speed type size is different than expected");
+#endif // C_ONLY_INTERFACE
+
+/**
+ * A tile of live traffic data.  The layout is:
+ *
+ * TileHeader (24 bytes)
+ * n x Speed entries (n x 2 bytes)
+ * IncidentsHeader (8 bytes)
+ * m x Incidents entries (buffer 0) (m x 8 bytes)
+ * m x Incidents entries (buffer 1) (m x 8 bytes)
+ *
+ * The Incidents are sparse - the TileHeader->incident_buffer_size specifies how
+ * much space is available in the tile.  The IncidentsHeader fields specify where
+ * incidents begin/end in this buffer.  This allows for new incident lists to be
+ * placed in the buffer, then the begin/end updated to atomicly move readers to
+ * a new location.
+ */
+#ifndef C_ONLY_INTERFACE
+class Tile {
+public:
+  Tile(char* tile_ptr)
+      : header{reinterpret_cast<TileHeader*>(tile_ptr)}, speeds{reinterpret_cast<volatile Speed*>(
+                                                             tile_ptr + sizeof(TileHeader))},
+        incident_count_0{reinterpret_cast<volatile uint32_t*>(
+            tile_ptr + sizeof(TileHeader) +
+            sizeof(Speed) * (tile_ptr == nullptr ? 0 : header->directed_edge_count))},
+        incident_count_1{reinterpret_cast<volatile uint32_t*>(
+            tile_ptr + sizeof(TileHeader) +
+            sizeof(Speed) * (tile_ptr == nullptr ? 0 : header->directed_edge_count) +
+            sizeof(*incident_count_0))},
+        incidents_0{reinterpret_cast<volatile Incident*>(
+            tile_ptr + sizeof(TileHeader) +
+            sizeof(Speed) * (tile_ptr == nullptr ? 0 : header->directed_edge_count) +
+            sizeof(*incident_count_0) * 2)},
+        incidents_1{reinterpret_cast<volatile Incident*>(
+            tile_ptr + sizeof(TileHeader) +
+            sizeof(Speed) * (tile_ptr == nullptr ? 0 : header->directed_edge_count) +
+            sizeof(*incident_count_0) * 2 +
+            (tile_ptr == nullptr ? 0 : header->incident_buffer_size) * sizeof(Incident))} {
+  }
+
+  Tile(const Tile& other) = default;
+  Tile& operator=(const Tile& other) = default;
+
+  const Speed getTrafficForDirectedEdge(const uint32_t directed_edge_offset) const {
+    if (header == nullptr)
+      return {};
+    assert(directed_edge_offset < header->directed_edge_count);
```

Comment:
we should throw here that way a running server will catch and return an error rather than crash and burn.